sphinx/tutorial/rose/furthertopics/upgrading-macro-development.rst

```diff
@@ -0,0 +1,568 @@
+   value will not be overwritten. In our case, we don't need a
+   ``rose-macro-remove.conf`` file.
+
+Go ahead and upgrade the app configuration to ``0.3``, as you did before.
+
+The ``rose-app.conf`` should now contain the new option, ``l_rudder_branch``.
+
+More Complex Upgrade Macros
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. TODO - link the upgrade macro API
+
+The Upgrade Macro API gives us quite a bit of power without having to write
+too much Python.
+
+For our ``1.0`` release, we want to add sailing gear - we need the number
+of ``misc_branches`` to be at least ``6``, and add a ``sail_canvas_sq_m``
+option.
+
+We may want to issue a warning for a deprecated option
+(``paddle_branches``) so that the user can decide whether to remove it.
+
+Create the file ``~/rose-meta/make-boat/1.0 directory/rose-meta.conf``
+and paste in the following configuration:
+
+.. code-block:: rose
+
+   [namelist:materials=hollow_tree_trunks]
+   compulsory=true
+   values=1
+
+   [namelist:materials=l_rudder_branch]
+   compulsory=true
+   type=logical
+
+   [namelist:materials=misc_branches]
+   compulsory=true
+   range=6:
+   type=integer
+
+   [namelist:materials=outrigger_tree_trunks]
+   compulsory=true
+   values=2
+
+   [namelist:materials=paddling_branches]
+   range=0:
+   type=integer
+   warn-if=True # Deprecated - real sailors don't use engines
+
+   [namelist:materials=sail_canvas_sq_m]
+   range=4:
+   type=real
+
+We need to write a macro that reflects these changes.
+
+We need to start with appending the following code to ``versions.py``:
+
+.. code-block:: python
+
+   class MyMoreComplexUpgradeMacro(rose.upgrade.MacroUpgrade):
+
+       """Upgrade from 0.3 (Amazing Ama) to 1.0 (Tremendous Trimaran)."""
+
+       BEFORE_TAG = "0.3"
+       AFTER_TAG = "1.0"
+
+       def upgrade(self, config, meta_config=None):
+           """Upgrade the boat!"""
+           # Some code doing something to config goes here.
+           return config, self.reports
+
+We already know how to add an option, so replace
+``# Some code going here...`` with
+``self.add_setting(config, ["namelist:materials", "sail_canvas_sq_m"], "5")``
+
+To perform the check/change in the number of ``misc_branches``, we can
+insert the following lines after the one we just added:
+
+.. code-block:: python
+
+           branch_num = self.get_setting_value(
+                  config, ["namelist:materials", "misc_branches"])
+           if branch_num.isdigit() and float(branch_num) < 6:
+               self.change_setting_value(
+                        config, ["namelist:materials", "misc_branches"], "6")
+
+This extracts the value of ``misc_branches`` (as a string!) and if the
+value represents a positive integer that is less than 6, changes it to
+``"6"``. It's good practice to guard against the possibility that a user
+might have set the value to a non-integer representation like ``'many'``
+- if we don't do this, the macro may crash out when running things like
+``float``.
```

Comment:
'like **a float (i.e. decimal)**' I assume is meant?

Reply:
Also perhaps use a '.. admonition:: Good Practice' box for this, as used e.g. in the 'rose-arch' file.

Reply:
This is referring to the python builtin ``float``.